File: android_mobile/src/main/java/com/alexstyl/specialdates/images/ImageLoader.java
package com.alexstyl.specialdates.images;

import android.graphics.Bitmap;
import android.support.annotation.Px;
import android.widget.ImageView;

import com.alexstyl.specialdates.Optional;

import java.net.URI;

public interface ImageLoader {

    Request load(URI imagePath);

    interface Request {

        FixedSizeRequest withSize(@Px int width, @Px int height);

        void into(ImageView imageView);

        Request asCircle();
    }

    interface FixedSizeRequest {
        Optional<Bitmap> synchronously();

        void into(ImageLoadedConsumer consumer);
    }

}

File: android_mobile/src/main/java/com/alexstyl/specialdates/images/ImageLoader.kt
package com.alexstyl.specialdates.images

import android.graphics.Bitmap
import android.support.annotation.Px
import android.widget.ImageView
import com.alexstyl.specialdates.Optional
import com.alexstyl.specialdates.contact.ImageURL

interface ImageLoader {

    fun load(imagePath: ImageURL): Request

    interface Request {

        fun withSize(@Px width: Int, @Px height: Int): FixedSizeRequest

        fun into(imageView: ImageView)

        fun asCircle(): Request
    }

    interface FixedSizeRequest {
        fun synchronously(): Optional<Bitmap>

        fun into(consumer: ImageLoadedConsumer)
    }

}
